aws-analytics-pinpoint/src/test/java/com/amplifyframework/analytics/pinpoint/AutoSessionTrackerTest.java

```diff
@@ -0,0 +1,185 @@
+
+import android.app.Activity;
+import android.app.Application;
+import android.os.Bundle;
+
+import com.amazonaws.mobileconnectors.pinpoint.analytics.AnalyticsClient;
+import com.amazonaws.mobileconnectors.pinpoint.analytics.SessionClient;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InOrder;
+import org.robolectric.RobolectricTestRunner;
+
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+
+/**
+ * Tests the {@link AutoSessionTracker}.
+ */
+@RunWith(RobolectricTestRunner.class)
+public final class AutoSessionTrackerTest {
+    private AnalyticsClient analyticsClient;
+    private SessionClient sessionClient;
+    private Application.ActivityLifecycleCallbacks callbacks;
+
+    /**
+     * Setup dependencies and object under test.
+     */
+    @Before
+    public void setup() {
+        // Dependencies
+        this.analyticsClient = mock(AnalyticsClient.class);
+        this.sessionClient = mock(SessionClient.class);
+
+        // Object under test
+        this.callbacks = new AutoSessionTracker(analyticsClient, sessionClient);
+    }
+
+    /**
+     * When the app is opened, a start session should be recorded.
+     */
+    @Test
+    public void sessionStartedWhenAppOpened() {
+        // Given: the launcher activity instance and bundle class instance.
+        Activity activity = mock(Activity.class);
+        Bundle bundle = mock(Bundle.class);
+
+        // When: the app is opened main activity goes through the following lifecycle states.
+        callbacks.onActivityCreated(activity, bundle);
+        callbacks.onActivityStarted(activity);
+        callbacks.onActivityResumed(activity);
+
+        // Then: Make sure that startSession was invoked on the session client.
+        verify(sessionClient).startSession();
+    }
+
+    /**
+     * When the app is started, user interacts with the app and presses the home button
+     * a start session and a stop session should be recorded in that order.
+     */
+    @Test
+    public void sessionStoppedWhenHomeButtonPressed() {
+        // Given: the activity that is currently in foreground.
+        Activity activity = mock(Activity.class);
+        Bundle bundle = mock(Bundle.class);
+
+        // When: the app goes to background, the activity in foreground is stopped.
+        // Activity is put in resume state when the app is started.
+        callbacks.onActivityCreated(activity, bundle);
+        callbacks.onActivityStarted(activity);
+        callbacks.onActivityResumed(activity);
+
+        // Activity stopped when home button is pressed.
+        callbacks.onActivityPaused(activity);
+        callbacks.onActivityStopped(activity);
+
+        // Then: Make sure startSession and stopSession are invoked in that order.
+        InOrder inOrder = inOrder(sessionClient);
```

Comment:
I've never seen this `InOrder`, but it looks quite useful. Nice find, and nice application.

Reply:
Came across while going through the documentation 😄